src/QsCompiler/Transformations/CallGraphWalker.cs

```diff
@@ -19,6 +18,243 @@ namespace Microsoft.Quantum.QsCompiler.Transformations
     using ExpressionKind = QsExpressionKind<TypedExpression, Identifier, ResolvedType>;
     using ResolvedTypeKind = QsTypeKind<ResolvedType, UserDefinedType, QsTypeParameter, CallableInformation>;
 
+    /// <summary>
+    /// Implementation of Johnson's algorithm for finding all simple cycles in a graph.
+    /// A simple cycle is one where no node repeats in the cycle until getting back
+    /// to the beginning of the cycle.
+    ///
+    /// The algorithm uses Tarjan's algorithm for finding all strongly-connected
+    /// components in a graph. A strongly-connected component, or SCC, is a subgraph
+    /// in which all nodes can reach all other nodes in the subgraph.
+    ///
+    /// The algorithm begins by using Tarjan's algorithm to find the SCCs of the graph.
+    /// Each SCC is processed for cycles, using the smallest-id node of the SCC as the
+    /// starting node. Then the starting node is removed from the graph, and Tarjan is used
+    /// again to find SCCs of the new graph, and the next SCC is processed. This repeats
+    /// until there are no more SCCs (and no more nodes) left in the graph.
+    ///
+    /// The algorithm has time complexity of O((n + e)(c + 1)) with n nodes, e edges,
+    /// and c elementary cycles.
+    ///
+    /// The starting point for this algorithm in this class is the GetAllCycles method.
+    /// </summary>
+    internal class JohnsonCycleFind
+    {
+        private Stack<(HashSet<int> SCC, int MinNode)> SccStack = new Stack<(HashSet<int> SCC, int MinNode)>();
+
+        /// <summary>
+        /// Johnson's algorithm for finding all cycles in a graph.
+        /// This returns a list of cycles, each represented as a list of nodes. Nodes
+        /// for this algorithm are integers. The cycles are guaranteed to not contain
+        /// any duplicates and the last node in the list is assumed to be connected
+        /// to the first.
+        /// </summary>
+        
+        public List<List<int>> GetAllCycles(Dictionary<int, List<int>> graph) =>
+            TarjanSCC(graph)
+                .OrderByDescending(x => x.MinNode)
+                .SelectMany(item =>
```

Comment:
>SelectMany [](start = 17, length = 10)

As per our email discussion, I would favor the old non-recursive approach.

Reply:
Moved to using the 'filter' implementation we discussed.